Add Swappy & Pre-Transformed Swapchain

darksylinc opened the pull request:
- Adds Swappy for Android for stable frame pacing
- Implements pre-transformed Swapchain so that Godot's compositor is in charge of rotating the screen instead of Android's compositor (performance optimization for phones that don't have HW rotator)

============================

The work was performed by collaboration of TheForge and Google. I am merely splitting it up into smaller PRs and cleaning it up.

Changes from original PR:

- Removed "display/window/frame_pacing/android/target_frame_rate" option to use Engine::get_max_fps instead.
- Target framerate can be changed at runtime using Engine::set_max_fps.
- Swappy is enabled by default.
- Added documentation.
- enable_auto_swap setting is replaced with swappy_mode.

___

- *Production edit: This closes https://github.com/godotengine/godot-proposals/issues/2351.*

darksylinc committed:
Add Swappy & Pre-Transformed Swapchain

- Adds Swappy for Android for stable frame pacing
- Implements pre-transformed Swapchain so that Godot's compositor is in
charge of rotating the screen instead of Android's compositor
(performance optimization for phones that don't have HW rotator)

============================

The work was performed by collaboration of TheForge and Google. I am
merely splitting it up into smaller PRs and cleaning it up.

Changes from original PR:

- Removed "display/window/frame_pacing/android/target_frame_rate" option
to use Engine::get_max_fps instead.
- Target framerate can be changed at runtime using Engine::set_max_fps.
- Swappy is enabled by default.
- Added documentation.
- enable_auto_swap setting is replaced with swappy_mode.

m4gr3d commented on platform/android/java/lib/src/org/godotengine/godot/GodotIO.java:
```diff
@@ -287,6 +294,21 @@ public int getScreenOrientation() {
 	}
 
+	public int getCurrentScreenRotation() {
+		int rotation = ((WindowManager)ContextCompat.getSystemService(activity.getBaseContext(), WindowManager.class)).getDefaultDisplay().getRotation();
+
```
Just use `activity.getSystemService(...)` to retrieve the `WindowManager`, not need to use the `ContextCompat` class.

m4gr3d commented on platform/android/java_godot_lib_jni.cpp:
```diff
@@ -83,7 +83,45 @@ static Vector3 gravity;
-static void _terminate(JNIEnv *env, bool p_restart = false) {
+// Thermal headroom support
+jclass godot_thermal_wrapper_class = nullptr;
+JNIEnv *godot_thermal_env = nullptr;
+jmethodID godot_thermal_wrapper_thermal_headroom_methodid = nullptr;
+jobject godot_thermal_wrapper_obj = nullptr;
+
+void GodotGetThermalHeadroom_initialize(JNIEnv *p_env) {
+	godot_thermal_env = p_env;
+	// Find the class
+	godot_thermal_wrapper_class = p_env->FindClass("org/godotengine/godot/GodotThermalWrapper$Companion");
+	if (godot_thermal_wrapper_class == nullptr) {
+		return;
+	}
+	// Find the method
+	godot_thermal_wrapper_thermal_headroom_methodid = p_env->GetMethodID(godot_thermal_wrapper_class,
+			"getThermalHeadRoom", "(I)F");
+	godot_thermal_wrapper_obj = godot_thermal_env->AllocObject(godot_thermal_wrapper_class);
+}
+
+void GodotGetThermalHeadroom_cleanup(JNIEnv *p_env) {
+	if (godot_thermal_wrapper_obj) {
+		godot_thermal_env->DeleteLocalRef(godot_thermal_wrapper_obj);
+	}
+}
+
+float GodotGetThermalHeadroom(int p_forecast_seconds) {
+	if (godot_thermal_wrapper_class == nullptr) {
+		return 0.0f;
+	}
+	if (godot_thermal_wrapper_thermal_headroom_methodid == nullptr) {
+		return 0.0f;
+	}
+
+	// Call the method
+	jfloat result = godot_thermal_env->CallFloatMethod(godot_thermal_wrapper_obj, godot_thermal_wrapper_thermal_headroom_methodid, p_forecast_seconds);
+	return result;
+}
+
+static void _terminate(JNIEnv *p_env, bool p_restart = false) {
```
Create a separate class and move that logic there.

m4gr3d commented on main/main.h:
```diff
@@ -69,6 +69,28 @@ class Main {
 	static const Vector<String> &get_forwardable_cli_arguments(CLIScope p_scope);
 #endif
 
+	// Thermal state support
+	enum ThermalState {
+		THERMAL_STATE_MIN = -2,
+		THERMAL_STATE_NOT_SUPPORTED = -2, // Current platform/device/build doesn't support ThermalState queries
+		THERMAL_STATE_ERROR = -1, // error acquiring thermal state
+
+		THERMAL_STATE_NONE = 0, // no state (perhaps device doesn't support it)
+		THERMAL_STATE_LIGHT = 1, // iOS Nominal
+		THERMAL_STATE_MODERATE = 2, // iOS Fair
+		THERMAL_STATE_SEVERE = 3, // iOS Serious
+		THERMAL_STATE_CRITICAL = 4, // iOS Critical
+		THERMAL_STATE_EMERGENCY = 5,
+		THERMAL_STATE_SHUTDOWN = 6,
+
+		THERMAL_STATE_MAX,
+	};
+	static ThermalState thermal_state;
+	static const char *get_therrmal_state_string(ThermalState p_thermalState);
+	static void update_thermal_state(ThermalState p_thermalState);
+	static ThermalState get_thermal_state();
+	static float get_thermal_headroom(int p_forecast_seconds);
```
Is there a reason we have this logic in `main` instead of its own class?

m4gr3d commented on platform/android/display_server_android.cpp:
```diff
@@ -216,6 +216,14 @@ DisplayServer::ScreenOrientation DisplayServerAndroid::screen_get_orientation(in
 
+int DisplayServerAndroid::screen_get_internal_current_rotation(int p_screen) const {
+	GodotIOJavaWrapper *godot_io_java = OS_Android::get_singleton()->get_godot_io_java();
+	ERR_FAIL_NULL_V(godot_io_java, SCREEN_LANDSCAPE);
```
This returns `SCREEN_LANDSCAPE` on failure, but the values returned by the java implementation are `0, 90, 180, 270` which don't map to the `ScreenOrientation` enum.

darksylinc replied:
Good point. SCREEN_LANDSCAPE mapped to 0 which was the intended value.
Anyway, changed to 0.

m4gr3d commented on servers/display_server.h:
```diff
@@ -367,6 +374,10 @@ class DisplayServer : public Object {
 	};
 
+private:
+	ScreenOrientation native_orientation = ScreenOrientation::SCREEN_LANDSCAPE;
```
What's the use of this field?

darksylinc replied:
Dead code. Removed. Thanks.

darksylinc commented on main/main.h:
```diff
@@ -69,6 +69,28 @@ class Main {
+	// Thermal state support
+	enum ThermalState {
+		THERMAL_STATE_MIN = -2,
+		THERMAL_STATE_NOT_SUPPORTED = -2, // Current platform/device/build doesn't support ThermalState queries
```
Intended.

I changed it to `THERMAL_STATE_NOT_SUPPORTED = THERMAL_STATE_MIN` to make it more explicit.

darksylinc commented on platform/android/java_godot_io_wrapper.h:
```diff
@@ -86,6 +87,8 @@ class GodotIOJavaWrapper {
 	void set_screen_orientation(int p_orient);
 	int get_screen_orientation();
+	int get_current_screen_rotation();
+	int get_native_screen_orientation();
```
Looks like a left over from TF that I missed from pruning (they handled a bunch of dead code I had to strip).

Thanks.

darksylinc commented:
A little bit of background from TheForge's PR. They introduced the following:

- Swappy, to improve frame pacing (I do not know about XR, but on regular devices Swappy makes a huge difference).
- Pre-transformed swapchain: This makes a tiny bit of performance difference on devices that do not have a HW rotator. For this to work Godot needs the return value of `get_current_screen_rotation`.
- Thermal code.

The most controversial change is likely the Thermal code:

- **It is not actually used, *except* for logging.**
   - Thermal events help when troubleshooting performance and pacing issues, because phones tend to throttle themselves when warm or hot. That's the main reason of its inclusion. Perhaps there could be a better use for it in the future; but were it not for its debugging usefulness, [it's borderline a solution looking for a problem](https://docs.godotengine.org/en/stable/contributing/development/best_practices_for_engine_contributors.html).
- It's not exposed to GDScript (should it? is there a use case of it?), but it is exposed in C++.
- I suspect TheForget didn't know how/where to place them other than the main activity (after all, we must start listening in `onCreate()`). TBH I didn't know what to do with it either.

darksylinc commented on platform/android/detect.py:
```diff
@@ -200,6 +203,8 @@ def configure(env: "SConsEnvironment"):
     env.Append(CPPDEFINES=["ANDROID_ENABLED", "UNIX_ENABLED"])
     env.Append(LIBS=["OpenSLES", "EGL", "android", "log", "z", "dl"])
+    env.Append(LIBS=["swappy_static"])
+    env.Append(LIBPATH=["java/editor/build/intermediates/stripped_native_libs/dev/out/lib/arm64-v8a"])
```
Now that CI failed I understand your question.
gitignore made me miss libswappy_static.a and it wasn't pushed. I just updated the PR with the missing files.

However I still need to check whether `java/editor/build/intermediates/stripped_native_libs/dev/out/lib/arm64-v8a` is needed. It sounds like it's not.

darksylinc replied:
Solved. That line wasn't needed.

darksylinc commented:
CI is blocking the PR on the basis `libswappy_static.a` is blocked from being pushed.

The traditional way of adding Swappy is by adding the library by downloading the [AGDK from Android](https://developer.android.com/games/downloads).

However the [AGDK is FOSS](https://android.googlesource.com/platform/frameworks/opt/gamesdk/) (Apache License), but it is a PITA to build from source code.

How do you want to proceed? Include *.a file as an exception? Build from source?

bruvzg commented:
> How do you want to proceed? Include *.a file as an exception? Build from source?

We are using separate repos for slow/hard to build libraries: [Mesa/NIR](https://github.com/godotengine/godot-nir-static), [MoltenVK](https://github.com/godotengine/moltenvk-osxcross) and [ANGLE](https://github.com/godotengine/godot-angle-static), it's build by from source, but main repo CI (and release build system) simply download release binaries (and there's build flags to override path to these).

darksylinc commented:
Ah I remember those repos.

How should I proceed? Do I prepare a repo, add it to github and then if it's Ok'ed it you push it to godotengine?
And SCons should always download (and cache) the lib when building for Android?

bruvzg commented:
> How should I proceed? Do I prepare a repo, add it to github and then if it's Ok'ed it you push it to godotengine?

That's what was done for these repos. When PR is merged, you can transfer repo to the `godotengine` organization.

> And SCons should always download (and cache) the lib when building for Android?

For the other libs, SCons is not downloading anything, there's a separate script which is [called by CI](https://github.com/godotengine/godot/blob/61598c5c88d95b96811d386cb20d714c35f4c6d7/.github/workflows/windows_builds.yml#L64-L65).

BastiaanOlij commented on core/config/project_settings.cpp:
```diff
@@ -1492,6 +1492,10 @@ ProjectSettings::ProjectSettings() {
 	// Keep the enum values in sync with the `DisplayServer::VSyncMode` enum.
 	custom_prop_info["display/window/vsync/vsync_mode"] = PropertyInfo(Variant::INT, "display/window/vsync/vsync_mode", PROPERTY_HINT_ENUM, "Disabled,Enabled,Adaptive,Mailbox");
+
+	GLOBAL_DEF("display/window/frame_pacing/android/enable_frame_pacing", true);
+	GLOBAL_DEF(PropertyInfo(Variant::INT, "display/window/frame_pacing/android/swappy_mode", PROPERTY_HINT_ENUM, "pipeline_forced_on,auto_fps_pipeline_forced_on,auto_fps_auto_pipeline"), 2);
```
In theory for OpenXR the swapchain on Android should never be created so I'm assuming this logic will thus be completely ignored.
But it's probably smart to disable it in this scenario just to be safe.

dsnopek replied:
Hm, looking in `rendering_device_driver_vulkan.cpp`, I think some amount of the Swappy-related code would run, even with OpenXR. It's `VulkanHooks` that allows OpenXR to take over the Vulkan setup, but it appears like some of the setup code would still happen.

So, I agree, it would be good to force Swappy to be disabled if OpenXR is being used.

darksylinc replied:
Is there an easy/quick way I can check we're running on OpenXR so I can simply skip Swappy code?

dsnopek replied:
An indirect way could be checking if `VulkanHooks::get_singleton() != nullptr`? I don't know if this is a safe assumption, but we could assume that if something is taking over creation of the Vulkan device, then we don't want to use Swappy?

There's more direct checks we could do, but it would mean code under `drivers/vulkan` interacting with code from `modules/openxr` which isn't normally allowed. Or, I suppose we could look at project settings, but that isn't ideal either.

Thoughts? cc @m4gr3d @BastiaanOlij

darksylinc replied:
I went for your solution `VulkanHooks::get_singleton() != nullptr` because it felt like the cautious thing to do (avoid Swappy) when we're not controlling device creation and possibly swapchain and presentation.

If someone wants to take this more fine grained like adding VulkanHooks::get_singleton()->allow_swappy() or whatever, that can be done in another PR.

darksylinc commented:
OK after some fighting with everything, I managed to put up https://github.com/darksylinc/godot-swappy/ , have the CI download the prebuilt version from there and now CI is passing.

I will now apply/fix some of the suggestions; but I have a question:

This change undoubtedly changes how Android is built from source, from a regular user perspective, i.e. a user trying to build from source will fail to link because of missing swappy lib.

Thus I have the following questions:

1. Should Scons detect if Swappy is there and disable Swappy supports if it's not included?
   - Or simply refuse to start building and tell the user to download Swappy? (like it asks for NIR if D3D12 support is enabled)
2. Do I have to submit a separate PR to [add the step of downloading Swappy in the documentation](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_android.html)?.
3. Should Scons instead try to detect if Swappy is there, and if it's not, download it from the internet?

clayjohn commented:
@darksylinc My preference is for 1. 

I think the simplest thing for you to do will be to make swappy a build option (like ANGLE). By default, we would build official builds with swappy=yes. But end users would default to swappy=no.

darksylinc commented:
> My preference is for 1.
> 
> I think the simplest thing for you to do will be to make swappy a build option (like ANGLE). By default, we would
> build official builds with swappy=yes. But end users would default to swappy=no.

Done!

I added `disable_swappy` CLI setting. It defaults to False. Thus users must compile with `disable_swappy=True` if they don't have Swappy and want to ignore the dependency.

1. If users try to compile without Swappy and without `disable_swappy=True` they'll be met with an error explaining they must either download Swappy and unpack it, or compile with `disable_swappy=True`
2. If users try to compile without Swappy and with `disable_swappy=True` they'll be met with a warning that Swappy was not found and they should download Swappy and unpack it. But compilation will continue and succeed.

darksylinc commented on platform/android/java/lib/src/org/godotengine/godot/GodotIO.java:
```diff
@@ -287,6 +294,21 @@ public int getScreenOrientation() {
 		}
 	}
 
+	public int getCurrentScreenRotation() {
```
I renamed it to `getInternalCurrentScreenRotation()` instead because it mimics what we did for `DisplayServer::screen_get_internal_current_rotation` (since it's literally the reason the Java version was added).

darksylinc commented on platform/android/java/lib/src/org/godotengine/godot/GodotActivity.kt:
```diff
@@ -41,6 +41,26 @@ import androidx.fragment.app.FragmentActivity
+		}
+	}
+}
+
```
I moved Thermal changes out of this PR.

I don't think they're ready for inclusion. It's too Android-centric (ignores iOS) and invades main code instead of living in its own class; and it's not clearly defined what its purpose is.

Also, it's stalling Swappy & Pre-Transformed Swapchains which are much higher priority.

This should be ready for review again.

Although... I cannot test it on Android because of #97341

m4gr3d commented on platform/android/detect.py:
```diff
@@ -176,19 +185,35 @@ def configure(env: "SConsEnvironment"):
     if env["arch"] == "x86_32":
         # The NDK adds this if targeting API < 24, so we can drop it when Godot targets it at least
         env.Append(CCFLAGS=["-mstackrealign"])
+        env.Append(LIBPATH=["../../thirdparty/swappy-frame-pacing/x86"])
```
Should this also have the `has_swappy:` check?

darksylinc replied:
Yes. Fixed.

darksylinc commented:
Hi! 

> > @darksylinc My preference is for 1.
> > I think the simplest thing for you to do will be to make swappy a build option (like ANGLE). By default, we would build official builds with swappy=yes. But end users would default to swappy=no.
> 
> @darksylinc Based on @clayjohn's comment, shouldn't `disable_swappy=True` be the default behavior?
> 
> Unless I'm misunderstanding the setup, I'd lean to have that being the default behavior as well so that Android builds from source are not 'broken' after this PR is merged. As Clay mentioned though, we should update https://github.com/godotengine/godot-build-scripts/blob/main/build-android/build.sh to enable building the official builds with swappy enabled.

I initially thought the same, but now I strongly recommend that `disable_swappy=False` be the default behavior, due to the following:

1. Unlike D3D12 and other similar features, it is not obvious when Godot was compiled without them. No D3D12? Oh that's obvious. The project will fail to launch with D3D12, the option in the Editor won't be available. It's immediately obvious it was not included. But Swappy? It's too easy not to notice it. Android's behavior will be exactly the same except for poor frame pacing (i.e. stutter), which psychologically can be misattributed to a number of things: slow phone, device temperature / throttling behaviors, unoptimized Godot code,  unoptimized user code,  poor LOD settings, driver issues, poor CPU/kernel scheduling, power saving behavior, apps in the background, etc. Even with Swappy enabled the things I mentioned can cause poor frame pacing (like in any other platform, though throttling & CPU scheduling tend to be much bigger issues on Android) which can make Swappy look like snake oil. But without Swappy stutter will happen even when everything else is not a problem. All the user will see is: The Editor launches. The project exports. The game runs on all supported devices. There's no apparent indication that something's wrong because the user compiled without Swappy.
2. I modified the Scons files so that users are immediately told of what happened (rather than having a bizarre clang compile error with no obvious way of how to fix) and telling them how to fix it (either compile with `disable_swappy=False` or download the dependency and unpack it).
3. Due to point 1, by "shoving" it into the user's build it becomes conscious decisions: basically I want to say "You're building for Android. Without Swappy. Are you sure?". Perhaps we should print the explanation from point 1 point in SCons. Explain far more thoroughly to the one building from source why Swappy is important.
4. This issue can accidentally creep into official builds. Failing to include `disable_swappy=False` into official build scripts CI can stay unnoticed for a long time. In the future, any new maintainer who needs to refactor the build scripts can incorrectly think `disable_swappy=False` is not needed or miss it by mistake.

darksylinc commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -2681,6 +2752,14 @@ void RenderingDeviceDriverVulkan::_swap_chain_release(SwapChain *swap_chain) {
 	if (swap_chain->vk_swapchain != VK_NULL_HANDLE) {
+#if defined(SWAPPY_FRAME_PACING_ENABLED)
+		if (swappy_frame_pacer_enable) {
+			// Swappy has a bug where the ANativeWindow will be leaked if we call
```
1. I already reported it to them, I haven't seen the fix yet. But it's been a while since the last time I checked their public source code.
2. It's still a good practice to do that, Swappy bug or not. Furthermore Swappy could end up "solving it" just change their API semantics to force the user to first destroy the Swapchain. The comment is basically explaining: Do NOT remove that call, Swappy won't do it for you (IMHO it's a Swappy bug nonetheless).

darksylinc commented:
> Also, the setting should probably be `swappy=yes|no` to match how other settings are worded (in a positive fashion, e.g. `vulkan=yes|no`).

I absolutely agree. It's just that I saw other `disable_*` settings and thought that was the norm.

m4gr3d commented:
> Unlike D3D12 and other similar features, it is not obvious when Godot was compiled without them. No D3D12? Oh that's obvious. The project will fail to launch with D3D12, the option in the Editor won't be available. It's immediately obvious it was not included. But Swappy? It's too easy not to notice it. Android's behavior will be exactly the same except for poor frame pacing (i.e. stutter), which psychologically can be misattributed to a number of things: slow phone, device temperature / throttling behaviors, unoptimized Godot code, unoptimized user code, poor LOD settings, driver issues, poor CPU/kernel scheduling, power saving behavior, apps in the background, etc. Even with Swappy enabled the things I mentioned can cause poor frame pacing (like in any other platform, though throttling & CPU scheduling tend to be much bigger issues on Android) which can make Swappy look like snake oil. But without Swappy stutter will happen even when everything else is not a problem. All the user will see is: The Editor launches. The project exports. The game runs on all supported devices. There's no apparent indication that something's wrong because the user compiled without Swappy.

No regular users compile Godot from source on their own, they'll get access to the official builds which will have swappy enabled. 
Whereas this will be a blocker for folks that actually build the editor from source (and an annoyance for folks building from source that don't need / use swappy (e.g: OpenXR)) and unfortunately not everyone stay up-to-date with github PRs, so this will lead to an influx of build 'issues'.

>  It's immediately obvious it was not included. But Swappy? It's too easy not to notice it.

Let's put the swappy project settings under a `#if defined(SWAPPY_FRAME_PACING_ENABLED)`. This way, if swappy is not included in the build, the project settings will be missing which will be a clear signal.

> This issue can accidentally creep into official builds. Failing to include disable_swappy=False into official build scripts CI can stay unnoticed for a long time. In the future, any new maintainer who needs to refactor the build scripts can incorrectly think disable_swappy=False is not needed or miss it by mistake.

Switching to a 'positive' setting should help here a bit, but this is a good callout that we should have a mean (preferably automatic) to validate whether the behavior is enabled or not.
Are there any suggestions in the swappy documentation of tests that could be used to validate swappy's impact when it's enabled, and thus that we could use to detect whenever the swappy implementation is accidentally disabled or broken?

m4gr3d commented on core/config/project_settings.cpp:
```diff
@@ -1485,6 +1485,10 @@ ProjectSettings::ProjectSettings() {
 	// Keep the enum values in sync with the `DisplayServer::VSyncMode` enum.
 	custom_prop_info["display/window/vsync/vsync_mode"] = PropertyInfo(Variant::INT, "display/window/vsync/vsync_mode", PROPERTY_HINT_ENUM, "Disabled,Enabled,Adaptive,Mailbox");
+
+	GLOBAL_DEF("display/window/frame_pacing/android/enable_frame_pacing", true);
+	GLOBAL_DEF(PropertyInfo(Variant::INT, "display/window/frame_pacing/android/swappy_mode", PROPERTY_HINT_ENUM, "pipeline_forced_on,auto_fps_pipeline_forced_on,auto_fps_auto_pipeline"), 2);
```
Follow-up comment to place this under a `#if defined(SWAPPY_FRAME_PACING_ENABLED)` guard so the settings are only available when swappy is compiled into the build.

darksylinc replied:
Done.

darksylinc replied:
CI is unhappy with this because it sees that enable_frame_pacing should not exist and tags the problem with ProjectSettings.xml

See https://github.com/godotengine/godot/actions/runs/11428066052/job/31792909945

clayjohn replied:
Project settings should be exposed regardless of the build setting to avoid exactly this error. But setting the project setting should be a no-op, or at most print a warning that the engine is compiled without support for that feature

darksylinc commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -533,6 +543,38 @@ Error RenderingDeviceDriverVulkan::_initialize_device_extensions() {
+		// Determine number of extensions required by Swappy frame pacer
+		SwappyVk_determineDeviceExtensions(physical_device, device_extension_count, device_extensions.ptr(), &swappy_required_extensions_count, nullptr);
+
+		if (swappy_required_extensions_count < device_extension_count && swappy_required_extensions_count < device_extension_count) {
```
Oops! Thanks. Fixed.

darksylinc commented:
OK I've done some changes to the PR:

1. `disable_swappy=no` is now `swappy=yes` (options shouldn't be expressive as negatives).
2. Expanded on the warning when Swappy is not found so that we explain why it is important (and not just say it is important / recommended).
3. Swappy is disabled by default BUT if "production=yes" is found, it defaults to enabled. I hope this is a reasonable compromise. Production builds should be held a higher standard, even if that occasionally breaks other build scripts.
4. Other misc / minor fixes pointed out by reviewers of the PR.

> cc @Calinou I wonder if we can update the benchmarks to take swappy into account on Android?

I don't know how to implement that. I mean in the sense we can do:

```cpp
if( production && !swappy )
   fail_tests();
```

But this should break other people's automated builds, whether it's at build time or at benchmark time.

If by "benchmark", you mean actually testing on the device and check framerate variance, then yes. Any benchmark should check:

``cpp
double avg_fps, std_dev_fps;

diff = abs( monitor_fps - avg_fps );

if( diff / monitor_fps > threshold )
   fail();

if( std_dev_fps > std_dev_threshold )
   fail();
```

And then make sure the thresholds are low enough to fail without Swappy but high enough pass with Swappy.

clayjohn reviewed (commented):
Overall the rendering changes look good to me. I'm also happy with the way it is exposed now. I'm happy to merge this once CI is passing. To do that, you will have to remove the #if defined` that was added around the project setting. While it is slightly awkward to have a project setting for a feature that may not be present, its better to have documentation for the feature. The only people who are missing swappy will be people who build themselves anyway, and they will get a big warning that they should be building with swappy

m4gr3d commented on platform/android/detect.py:
```diff
@@ -176,19 +185,35 @@ def configure(env: "SConsEnvironment"):
+    has_swappy = detect_swappy()
+    if not has_swappy:
+        print_warning(
+            "Swappy Frame Pacing not detected! It is strongly recommended you download it from https://github.com/darksylinc/godot-swappy/releases and extract it so that thirdparty/swappy-frame-pacing/arm64-v8a/libswappy_static.a can be found.\n"
```
> thirdparty/swappy-frame-pacing/arm64-v8a/libswappy_static.a

This should probably be generic since the architecture directory will change based on the arch. For example:
`thirdparty/swappy-frame-pacing/<arch_dir>/libswappy_static.a`

darksylinc replied:
This is more a social thing, which I'm not fully satisfied with the current wording (nor with all the alternatives I can think of).

Here's the rationale let's assume a very, very dumb user; I always assume worst case and then walk to the second-best case which is a smart user (best case is user starts reading the source code and figures it out, but that's an outlier):

**It says:** `thirdparty/swappy-frame-pacing/<arch_dir>/libswappy_static.a`
- User may reply "there is no <arch_dir> folder!!! What do I do???"
- Smart people will not be certain what folders are supposed to either. Just cross fingers they uncompressed it correctly.

**It says:** `thirdparty/swappy-frame-pacing/arm64-v8a/libswappy_static.a`
- User may reply "just arm64-v8a folder? But I am compiling for x86..."
- Smart people will figure out at least arm64-v8a is correct, and see the other folders and cross fingers they have all the platforms included.


**It says:** `thirdparty/swappy-frame-pacing/arm64-v8a/libswappy_static.a` + `thirdparty/swappy-frame-pacing/x86/libswappy_static.a` + (basically list all platforms)
- User will not read it. It was too long.
- Smart people will have no confusion.

There's basically no solution that will satisfy all cases. I think I prefer to list all platforms rather than use `<arch_dir>`. The goal is that when something goes wrong, we give enough information for users in hopes of **minimizing the amount people that will reach to tech support** (e.g. leave a ticket here, post a forum post in community channels, etc). And if they do reach support and end up being asked to upload the error message, the smart people will figure out what the user needs to do.

darksylinc replied:
Pushed new version with all 4 variations.

m4gr3d commented on platform/android/detect.py:
```diff
@@ -176,19 +185,35 @@ def configure(env: "SConsEnvironment"):
+    if not has_swappy:
+        print_warning(
+            "Swappy Frame Pacing not detected! It is strongly recommended you download it from https://github.com/darksylinc/godot-swappy/releases and extract it so that thirdparty/swappy-frame-pacing/arm64-v8a/libswappy_static.a can be found.\n"
+            + "Without Swappy, Godot apps on Android will inevitable suffer stutter and struggle to keep consistent 30/60/90/120 fps. Though Swappy cannot guarantee your app will be stutter-free, not having Swappy will guarantee there will be stutter even on the best phones and the most simple of scenes."
```
Seems a bit alarmist.. This paragraph is basically asserting that up until now, all Godot Android apps / games have been riddled with stutters.. is that the case?

If not, I'd recommend removing this section, and maybe adding that swappy will improve game performance to the previous paragraph.

darksylinc replied:
> all Godot Android apps / games have been riddled with stutters.. is that the case?

Yes. It's really bad.

m4gr3d commented on platform/android/detect.py:
```diff
@@ -176,19 +185,35 @@ def configure(env: "SConsEnvironment"):
     if get_min_sdk_version(env["ndk_platform"]) >= 24:
         env.Append(CPPDEFINES=[("_FILE_OFFSET_BITS", 64)])
 
     if env["arch"] == "x86_32":
```
Do we need an additional case to cover `x86_64`?

darksylinc replied:
Aahhh!!! I was confused by something and for some reason thought Godot did not support x86_64.
Thanks. Fixed.

m4gr3d reviewed (approved):
Add some minor comments to address, but otherwise it looks good!

AThousandShips commented on doc/classes/ProjectSettings.xml:
```diff
@@ -816,6 +816,17 @@
 		</member>
+		<member name="display/window/frame_pacing/android/enable_frame_pacing" type="bool" setter="" getter="" default="true">
+			Enable Swappy for stable frame pacing on Android. Highly recommended.
+			Note: This option will be forced off when using OpenXR.
```
```suggestion
			[b]Note:[/b] This option will be forced off when using OpenXR.
```

darksylinc commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -532,6 +539,38 @@ Error RenderingDeviceDriverVulkan::_initialize_device_extensions() {
+			free(swappy_required_extensions);
+		}
+	}
+
```
**Won't fix:** Newlines are strongly recommended to be part of the end of file for text files. See this [Stack Exchange answer](https://unix.stackexchange.com/a/18789/350793). My Editors add it automatically and keeps coming back.

In the case of C (not C++), a newline after EOF [is even mandatory](https://gcc.gnu.org/legacy-ml/gcc/2003-11/msg01568.html) per spec.

dsnopek replied:
This isn't the end of the file - it looks like this is just a space before an `#endif`

darksylinc replied:
OK that was weird. I looked and it was at the EOF.
WHAT SORCERY IS THIS???

Yeah I can fix that.

darksylinc replied:
Done.

clayjohn reviewed (approved):
Lets goooooooo! This has been through a very rigorous review process and its time to finally merge it. 

Everything looks great! 

Our one followup item will be to transfer https://github.com/darksylinc/godot-swappy/releases to the GodotEngine org so that we are hosting it. But we need to wait for akien-mga to do that. So let's merge as-is and then make the change later.

clayjohn commented:
Thank you!!!!

Alex2782 commented:
I first tried with "Source code", there were only the README and LICENSE files in there. The correct file is "godot-swappy.7z". A more detailed description might be good later.

![Screenshot 2024-10-29 214120](https://github.com/user-attachments/assets/964a0faa-ea5e-4e52-ba87-ec37aa0cce97)

```
        print_warning(
            "Swappy Frame Pacing not detected! It is strongly recommended you download it from https://github.com/darksylinc/godot-swappy/releases and extract it so that the following files can be found:\n"
            + " thirdparty/swappy-frame-pacing/arm64-v8a/libswappy_static.a\n"
            + " thirdparty/swappy-frame-pacing/armeabi-v7a/libswappy_static.a\n"
            + " thirdparty/swappy-frame-pacing/x86/libswappy_static.a\n"
            + " thirdparty/swappy-frame-pacing/x86_64/libswappy_static.a\n"
            + "Without Swappy, Godot apps on Android will inevitable suffer stutter and struggle to keep consistent 30/60/90/120 fps. Though Swappy cannot guarantee your app will be stutter-free, not having Swappy will guarantee there will be stutter even on the best phones and the most simple of scenes."
```

Alex2782 commented:
Does anyone know which Vulkan extensions are necessary for this?

`Samsung Galaxy Tab S6 lite.` Cannot initialize "VULKAN". 
https://forum.godotengine.org/t/code-keeps-duplicating-when-i-click-it/87411/12

Seems to work on my Samsung Tab S7. 
But the splash screen is rotated 180° when Vulkan renderer (Forward+ / Mobile) is used.


```cpp
#if defined(SWAPPY_FRAME_PACING_ENABLED)
	if (swappy_frame_pacer_enable) {
		char **swappy_required_extensions;
		uint32_t swappy_required_extensions_count = 0;
		// Determine number of extensions required by Swappy frame pacer.
		SwappyVk_determineDeviceExtensions(physical_device, device_extension_count, device_extensions.ptr(), &swappy_required_extensions_count, nullptr);

```

darksylinc commented:
Hi @Alex2782 !

Are you saying you've nailed down the problem to this specific PR?

Assuming the answer is yes, there's two things going on Swappy and Pre-Transformed swapchain.:

1. Swappy requires the extension `VK_GOOGLE_display_timing`. If it's not available, Swappy is turned off (or at least _should be_ turned off). It's more likely though, that `VK_GOOGLE_display_timing` is just buggy on your drivers and should be blocklisted.
2. The splash screen issue seems to be a silly bug with pre-transformed swapchains that we missed (Godot is now in charge of rotating the screen instead of the OS). Sometimes the problem is with Android which reported the wrong orientation and Godot is now stuck processing and can't query again for the orientation.

Alex2782 commented:
I will check it more closely `swappy=yes/no` on Firebase Test Lab to see whether and how many devices are affected.

m4gr3d commented:
@darksylinc I'm also seeing the inverted splash screen issue. I observed it on a debug build of the Android editor with swappy disabled running on Samsung Galaxy tab S8 using [Kenney's FPS starter sample](https://godotengine.org/asset-library/asset/2208).

Do you think that can be addressed before the dev4 release?

Alex2782 commented:
On Firebase Test Lab I couldn't find any differences between `swappy=yes` and `swappy=no` (with custom splash screen [Kenney's FPS starter sample](https://godotengine.org/asset-library/asset/2208), but with `Godot 4.3 stable` it is correct)

However, the user [kibby_Nibb](https://forum.godotengine.org/u/kibby_Nibb/summary) says that with `swappy=no` the Android Godot editor no longer shows Vulkan errors. I asked about the project if it could be shared.

darksylinc commented:
Hi!

I don't know about the schedule but as for the problem itself, luckily there are not many things that can go wrong:

After this PR Godot sets:

```cpp
VkSurfaceTransformFlagBitsKHR surface_transform_bits = surface_capabilities.currentTransform;
```

Which is basically saying "we're in charge of rotating the final image, not the OS".

Afterwards to accomplish this we added the following:

```cpp
blit.push_constant.rotation_cos = Math::cos(screen_rotation);
blit.push_constant.rotation_sin = Math::sin(screen_rotation);
```

Which should be self explanatory: The final blit done by the compositor performs the rotation in the pixel shader.


## What can go wrong?

1. The splash is not using the blit pass. I was certain everything Godot renders to the screen goes through that final blit pass from `RendererCompositorRD::blit_render_targets_to_screen`. If we somehow missed a render path that does not use it, then the screen won't be properly rotated.
   - I don't this is it. I downloaded the project you pointed to and replaced `DisplayServer::get_singleton()->screen_get_internal_current_rotation()` for a hardcoded "180" on PC, and the splash was inverted. Which means the blit pass must be in use.
2. `DisplayServer::get_singleton()->screen_get_internal_current_rotation()` is not returning the proper value. This might be some Android shenanigan where the value the OS returns is wrong up until a few seconds or presentations; but the splash screen never gets to see the correct value.
3. Android returned the wrong value in `surface_capabilities.currentTransform` on boot and will soon be corrected as soon as Godot catches the `VK_SUBOPTIMAL_KHR` "error" during presentation so that we recreate the swapchain and receive the right `surface_capabilities.currentTransform` .
4. Somehow the values used to send to the blit pass shader are uninitialized or the data we're pulling from is 1 frame behind (i.e. off by 1) and since nothing has been rendered, it's uninitialized. This should not happen because we're using push constants. But who knows.

As you noticed, Swappy has nothing to do with the inverted splash.

My money is on either 2 or 3, which would be Android being... Android. If so, the solution perhaps may be to try to render a black screen for a single frame, then the splash. Or to try to detect this situation and see how Android is behaving to detect the situation and correct it.

m4gr3d commented:
@darksylinc On the Quest, the entire editor is inverted 180 all the time.
 
I think we may need to take into account that not all Android devices (or forks of Android) will return the correct value. Is it possible to have a fallback where we let the OS be in charge of rotating the final image?

m4gr3d commented:
> @darksylinc On the Quest, the entire editor is inverted 180 all the time.
> 
> I think we may need to take into account that not all Android devices (or forks of Android) will return the correct value. Is it possible to have a fallback where we let the OS be in charge of rotating the final image?

Or are we properly detecting the rotation changes? 
In Google's documentation, I don't see references to `Display#getRotation()`; instead this is what they're using: https://developer.android.com/games/optimize/vulkan-prerotation#detect_device_orientation_changes

darksylinc commented:
OK I found the source of the problem. I'll submit a PR shortly with the fix.

> On the Quest, the entire editor is inverted 180 all the time.

Mmmm I think we are not the ones creating the swapchain on the Quest, and perhaps that's the problem.
However I first want to see if the PR's fix will fix the problem on the Quest.

If not, then yes, we'll have to disable the feature for the Quest, since it's a special case.

> I think we may need to take into account that not all Android devices (or forks of Android) will return the correct value. Is it possible to have a fallback where we let the OS be in charge of rotating the final image?

I think I made myself not clear: I meant that getRotation() may _**temporarily**_ return the wrong value. This is normal and expected (you may have experienced as a user that sometimes you tilt the device 90° and an app rotates -90°; then 180°).

If `getRotation()` always returned the wrong value, no Android app would be running correctly. I know Android is bad, but it is not THAT broken :rofl: .

Anyway, this is the source of the problem but I found an easy solution, that makes things simpler too. I'll elaborate on it on the PR.

clayjohn commented:
With respect to the splash screen, it is drawn by `RendererCompositorRD::set_boot_image` We probably need matching code in their to handle the rotation.

darksylinc commented:
PR submitted to fix the "Splash is upside down / 180°" bug: https://github.com/godotengine/godot/pull/98709

It may also fix the Quest issue too (I don't know).

DarioSamo commented:
I suspect this PR has added a regression when it concerns the handling of the extent size.

https://github.com/godotengine/godot/blob/1bffd6c73b44b85e5889f54e14b2193940cf5bb1/drivers/vulkan/rendering_device_driver_vulkan.cpp#L2888-L2912

VkSwapchainCreateInfoKHR now exclusively uses native_display_size, which is assigned from the value of surface_capabilities.currentExtent. However, right below the code that does this, you can see there was handling for the case that this returns UINT32_MAX.

https://github.com/godotengine/godot/blob/1bffd6c73b44b85e5889f54e14b2193940cf5bb1/drivers/vulkan/rendering_device_driver_vulkan.cpp#L2903C2-L2907

This handling is now gone and the UINT32_MAX value can go straight to the VkSwapchainCreateInfoKHR without being validated.

Since the logic for this was altered on this PR, I imagine you'll want to take a second look to see at what the intended handling was there. I suspect issues like [this](https://github.com/godotengine/godot/issues/98811) are happening due to this change being recent.
